src/builders/thermodynamic.jl

```diff
@@ -0,0 +1,258 @@
+Effectively this assumes an aqueous environment at constant pH is used.
+
+`constant_concentrations` is used to fix the concentrations of certain
+metabolites (such as CO₂). `concentration_ratios` is used to specify additional
+constraints on metabolite pair concentrations (typically, this is done with
+various cofactors such as the ATP/ADP ratio. For example, you can fix the
+concentration of ATP to be always 5× higher than of ADP by specifying
+`Dict(("ATP","ADP") => 5.0)`
+
+`concentration_lb` and `concentration_ub` set default concentration bounds, in M
+by default.
+
+`T` and `R` can be specified in the corresponding units; defaults are K and
+kJ/K/mol.
+"""
+function build_max_min_driving_force_model(
+    model::A.AbstractFBCModel,
+    reaction_standard_gibbs_free_energies::Dict{String,Float64};
+    flux_solution = Dict{String,Float64}(),
+    proton_ids = ["h_c", "h_e"],
+    water_ids = ["h2o_c", "h2o_e"],
+    concentration_lb = 1e-9,
+    concentration_ub = 100e-3,
+    T = 298.15, # Kelvin
+    R = 8.31446261815324e-3, # kJ/K/mol
+    ignore_reaction_ids = String[],
+)
+    # check if reactions that will be used in the model all have thermodynamic data, otherwise throw error.
+    # only these reactions will form part of the model.
+    rxns = filter(
+        x -> !(x in ignore_reaction_ids),
+        isempty(flux_solution) ? A.reactions(model) : collect(keys(flux_solution)),
+    )
+    all(in.(rxns, Ref(collect(keys(reaction_standard_gibbs_free_energies))))) || throw(
+        ArgumentError("""
+                      Not all reactions have thermodynamic data. 
+                      Either add the reactions with missing ΔG0s to ignore_reaction_ids, 
+                      or add the data to reaction_standard_gibbs_free_energies.
+                      """),
+    )
+
+    # import metabolite ids (if used), and reaction stoichiometries from an AbstractFBCModel
+    mets = Set(Symbol[])
+    for rid in rxns
+        for met in keys(A.reaction_stoichiometry(model, rid))
+            push!(mets, Symbol(met))
+        end
+    end
+    mets = [m for m in mets] # TODO: constrainttrees #12
+    stoi = A.stoichiometry(model)
+
+    # create thermodynamic variables
+    m = C.ConstraintTree(:max_min_driving_force^C.variable())
+    m +=
+        :log_metabolite_concentrations^C.variables(
+            keys = mets,
+            bounds = zip(log(concentration_lb), log(concentration_ub)),
+        )
+    m += :delta_G_reactions^C.variables(keys = Symbol.(rxns))
+
+    #= 
+    Build gibbs free energy relations
+    ΔG_rxn == ΔG0 + R * T * sum ( log_concentration_variable * stoichiometry_value )
+    =#
+    model_rxns = A.reactions(model)
+    model_mets = A.metabolites(model)
+    for rxn in rxns
+        met_idxs, stoich_coeffs = findnz(stoi[:, findfirst(==(rxn), model_rxns)])
+        met_ids = model_mets[met_idxs]
+
+        dG0 = reaction_standard_gibbs_free_energies[rxn]
+
+        m *=
```

Comment:
this mutates the thing million times (one per each reaction). Why run a loop when the dict can be constructed directly roughly as:

```
m *= :delta_g_reaction_equations ^ C.ConstraintTree(rid => .... for rid=rxns)
```
also please take care about rxns vs. rids naming

Reply:
aahh did not know :)

Reply:
but it is more clunky like that though